components/CategoriesBar/CategoriesBar.tsx

```diff
@@ -1,73 +1,66 @@
+    const containerRef = useRef<HTMLDivElement>(null);
     const { formatMessage } = useIntl();
+    const initialCategories = useCategories();
+    const [categories, setCategories] = useState({
+        nonOverflowing: initialCategories,
+        overflowing: [] as Category[],
+    });
 
-    const maxDisplayedCharacters = useCategoryCharacterLimit();
-
-    const filteredCategories = categories.filter(
-        (category) =>
-            category.stories_number > 0 && getCategoryHasTranslation(category, currentLocale),
-    );
-
-    const [visibleCategories, hiddenCategoriesCount] = useMemo(() => {
-        let characterCounter = 0;
-        let lastVisibleCategoryIndex = 0;
-
-        while (
-            characterCounter < maxDisplayedCharacters &&
-            lastVisibleCategoryIndex < filteredCategories.length
-        ) {
-            const { name } = getLocalizedCategoryData(
-                filteredCategories[lastVisibleCategoryIndex],
-                currentLocale,
-            );
-            characterCounter += name.length;
-
-            if (characterCounter < maxDisplayedCharacters || lastVisibleCategoryIndex === 0) {
-                lastVisibleCategoryIndex += 1;
-            }
+    useEffect(() => {
+        if (categories.overflowing.length > 0) {
+            const { nonOverflowing, overflowing } = categories;
+            setCategories({
+                ...categories,
+                nonOverflowing: differenceWith(nonOverflowing, overflowing, isEqual),
+            });
         }
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [categories.overflowing]);
 
-        return [
-            filteredCategories.slice(0, lastVisibleCategoryIndex),
-            filteredCategories.slice(lastVisibleCategoryIndex).length,
-        ];
-    }, [filteredCategories, currentLocale, maxDisplayedCharacters]);
-
-    if (!visibleCategories.length) {
+    if (!initialCategories.length) {
         return null;
     }
 
-    const hasMore = hiddenCategoriesCount > 0;
-
     return (
         <div className={styles.wrapper}>
-            <div className={styles.container}>
-                {visibleCategories.map((category) => (
-                    <CategoryLink key={category.id} category={category} />
+            <div className={styles.container} ref={containerRef}>
+                {categories.nonOverflowing.map((category) => (
+                    <CategoryLink
+                        key={category.id}
+                        category={category}
+                        containerRef={containerRef}
+                        onIsOverflowing={(isOverflowing) => {
+                            if (isOverflowing) {
+                                setCategories((prev) => ({
+                                    ...prev,
+                                    overflowing: [...prev.overflowing, category],
+                                }));
+                            }
+                        }}
```

Comment:
This looks overly complex to me. I don't think there's a need to check each category individually. If you already added `white-space: nowrap` to the category links, then you should be able to compare the `scrollWidth` of the container with its `innerWidth` (or `clientWidth`), removing the need for the complex logic in the `CategoryLink` component.

Reply:
But still, we can not be sure _how many and which items are overflowing_ so that we put them under the "more" link. That's why I thought individual check is necessary. I'll think more about this.